### PR TITLE
Elf Loader (Parser)

### LINKED ISSUE #9: Load and run ELF
Use [ELFIO](https://github.com/serge1/ELFIO)

- Load PIC ELF to memory of the process
- Validate ELF target arch, lack of `.rela.plt` & `.rela.dyn` sections
- Run the ELF from entry point
- UNIT TEST it!

### PR DESCRIPTION
**Pull request by c71n93**
Closes #9 

Adds:
- `ElfParser` class for parsing elf files.
- Implements `loadElf` method of `PhysMemBuilder` class
- Unit tests

This branch need to be squashed while merging into main!

### CODE REVIEW

File: include/besm-666/utils/elf-parser.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace besm::utl {
+
+class ElfParser final {
+public:
+    enum Requirements {
+        /// 64bit
+        FileClass = ELFIO::ELFCLASS64,
+        /// little-endian
+        Encoding = ELFIO::ELFDATA2LSB,
```
Comment (levBagryansky) [inline include/besm-666/utils/elf-parser.hpp:20]: @c71n93 why not ELFIO::LITTLE_ENDIAN, or ELFIO::LENDIAN?

File: include/besm-666/utils/elf-parser.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        /// RISC-V
+        Arch = ELFIO::EM_RISCV
```
Comment (levBagryansky) [inline include/besm-666/utils/elf-parser.hpp:22]: @c71n93 EM? Do not eat it)

File: include/besm-666/utils/elf-parser.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        LoadableSegment(LoadableSegment &&other);
+        LoadableSegment &operator=(LoadableSegment &&other);
+    };
+
+    ElfParser(const std::filesystem::path &elfPath);
+
+    const std::vector<LoadableSegment> &getLoadableSegments() &;
+
+private:
+    void checkRequirements();
```
Comment (levBagryansky) [inline include/besm-666/utils/elf-parser.hpp:40]: @c71n93 `void checkRequirements() const;`

File: test/elfgen/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+add_library(besm666_elfgen INTERFACE)
+target_include_directories(besm666_elfgen INTERFACE
+    .
+)
```
Comment (levBagryansky) [inline test/elfgen/CMakeLists.txt:4]: @c71n93 just emply library?

**Review by levBagryansky — CHANGES_REQUESTED** [review]

File: include/besm-666/utils/elf-parser.hpp (hunk 1 above)
Comment (c71n93) [reply include/besm-666/utils/elf-parser.hpp:20]: @levBagryansky  It's ELFIO internal constant

File: include/besm-666/utils/elf-parser.hpp (hunk 2 above)
Comment (c71n93) [reply include/besm-666/utils/elf-parser.hpp:22]: @levBagryansky It's ELFIO internal constant

File: test/elfgen/CMakeLists.txt (hunk 4 above)
Comment (c71n93) [reply test/elfgen/CMakeLists.txt:4]: @levBagryansky I added in `.` as `target_include_directories`

File: include/besm-666/util/elf-parser.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include <iostream>
+
+#include "besm-666/memory/phys-mem.hpp"
+#include "besm-666/riscv-types.hpp"
+#include "besm-666/util/non-copyable.hpp"
+#include "elfio/elfio.hpp"
+
+namespace besm::util {
+
+class ElfParser final {
```
Comment (mmamayka) [inline include/besm-666/util/elf-parser.hpp:14]: `ElfParser` contains non-copyable field `std::vector<LoadableSegment>`, so, please, mark the `ElfParser` class as non-copyable to make compiler error log more human readable.

File: include/besm-666/util/elf-parser.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+public:
+    enum Requirements {
```
Comment (mmamayka) [inline include/besm-666/util/elf-parser.hpp:16]: I suppose, `Requirements` should be public structure with `fileClass`, `encoding` and `arch` fields to make the `ElfParser` more customizable. The `Requirements` should be passed to ctor and checked after elf parsing.

Also, I suppose to add public static instance of `Requirements` structure, that describes RISCV standalone bare-metal ELF build used in out simulator.

File: include/besm-666/util/elf-parser.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    const std::vector<LoadableSegment> &getLoadableSegments() &;
+
+private:
+    void checkRequirements() const;
+
+    ELFIO::elfio reader_{};
+    std::vector<LoadableSegment> loadableSegments_;
+};
+
+class BadElf : public std::runtime_error {
```
Comment (mmamayka) [inline include/besm-666/util/elf-parser.hpp:46]: I suppose to rename this exception to `InvalidELFFormat` to throw it if ELF structure is invalid (`elfio` raises a error). And also to add `UnvailableELFRequirements` error to separate invalid elf format error and mismatch of requirements error.

File: include/besm-666/util/elf-parser.hpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+public:
+    BadElf(const std::string &msg);
```
Comment (mmamayka) [inline include/besm-666/util/elf-parser.hpp:48]: Please, add also `char const*` ctor

**Review by mmamayka — CHANGES_REQUESTED** [review]
Looks good, but should be polished a bit

**Review by levBagryansky — DISMISSED** [review]

**Review by mmamayka — APPROVED** [review]
LGBTQ++14

File: include/besm-666/util/elf-parser.hpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+     * a reference to it
+     */
+    virtual const std::vector<LoadableSegment> &getLoadableSegments() & = 0;
+
+    /**
+     * \brief creates instance of specific ElfParser that is inheritor of
+     * IElfParser
+     */
+    static std::unique_ptr<IElfParser>
+    createParser(const std::filesystem::path &elfPath);
```
Comment (levBagryansky) [inline include/besm-666/util/elf-parser.hpp:42]: @c71n93 I don't think it's ok that `interface` knows about its inheritors. Let's make this method belonging to `ElfParser` class

**Review by levBagryansky — CHANGES_REQUESTED** [review]
@c71n93 please check

File: include/besm-666/util/elf-parser.hpp (hunk 9 above)
Comment (levBagryansky) [reply include/besm-666/util/elf-parser.hpp:42]: @mmamayka WDYT?

**Review by levBagryansky — APPROVED** [review]

**Review by mmamayka — APPROVED** [review]
LGBTQ++17

**Pull request merged**